tracing/tracer.py

```diff
@@ -43,12 +63,48 @@ def active_trace(self) -> typing.Iterator[None]:
         finally:
             self.stop_trace()
 
-    def _reset_members(self) -> None:
-        """Resets the variables of the tracer."""
-        self.trace_data = pd.DataFrame(
-            columns=constants.TraceData.SCHEMA.keys()
-        ).astype(constants.TraceData.SCHEMA)
-        self.old_values_by_variable_by_function_name = {}
+    def _update_optimisations(self, fwm: FrameWithMetadata) -> None:
+        # Remove dead optimisations
+        while self.optimisation_stack:
+            top = self.optimisation_stack[-1]
+            if top.status() == TriggerStatus.EXITED:
+                logger.debug(
+                    f"Removing {self.optimisation_stack[-1].__class__.__name__} from optimisations"
+                )
+                self.optimisation_stack.pop()
+            else:
+                break
+
+        # Appending; only one optimisation at a time
+        # Check we do not trace somewhere we do not belong, e.g. Python's stdlib!
+        # NOTE: De Morgan - if no optimisations are on and we are in an unwanted path OR
+        # NOTE: if the newest optimisation is not a currently active Ignore and we are in an unwanted path
+        ignore = Ignore(fwm)
+        frame_path = pathlib.Path(fwm.co_filename)
+        if (
+            not self.optimisation_stack
+            or not isinstance(self.optimisation_stack[-1], Ignore)
+        ) and not frame_path.is_relative_to(self.project_dir):
+            logger.debug(f"Applying Ignore for {inspect.getframeinfo(fwm._frame)}")
+            self.optimisation_stack.append(ignore)
+            return
+
+        # Entering a loop for the first time
+        if (
+            not self.optimisation_stack
+            or not isinstance(self.optimisation_stack[-1], TypeStableLoop)
+        ) and fwm.is_for_loop():
+            logger.debug(
+                f"Applying TypeStableLoop for {inspect.getframeinfo(fwm._frame)}"
+            )
+            tsl = TypeStableLoop(fwm)
+            if not self.optimisation_stack or tsl != self.optimisation_stack[-1]:
+                self.optimisation_stack.append(tsl)
+                return
+
+    def _advance_optimisations(self, fwm: FrameWithMetadata) -> None:
+        for optimisation in self.optimisation_stack:
```

Comment:
Why is it a stack if it is used as a queue?

Reply:
It is used as a queue. Check `_update_optimisations`, which pops elements from the top.